File: Problems/Algorithms/1048. Longest String Chain/LongestStringChain.kt
class Solution {
    fun longestStrChain(words: Array<String>): Int {
        val n = words.size
        if (n == 1) return 1
        
        words.sortBy { it.length }
        
        val map: MutableMap<String, Int> = mutableMapOf()
        var ans = 0
        for (word in words) {
            var curr = 0
            for (i in 0..word.length-1) {
                val prev = word.substring(0, i) + word.substring(i+1)
                curr = maxOf(curr, map.getOrDefault(prev, 0)+1)
            }
            map.put(word, curr)
            ans = maxOf(ans, curr)
        }
        
        return ans
    }
}

File: Problems/Algorithms/1048. Longest String Chain/longest-string-chain.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), [](const string& a, const string& b) {
            return a.size() < b.size();
        });

        unordered_map<string, int> dp;
        int ans = 0;
        for (const auto& word : words) {
            dp[word] = 1;
            for (int i = 0; i < word.size(); i++) {
                string prev = word.substr(0, i) + word.substr(i+1);
                if (dp.find(prev) != dp.end()) {
                    dp[word] = max(dp[word], dp[prev] + 1);
                }
            }

            ans = max(ans, dp[word]);
        }

        return ans;
    }
};
